### ISSUE TITLE
r : Install R MPI library with OpenMPI ERROR, YOUR CLUSTER IS NOT READY YET

### ISSUE THREAD

**Issue by Charlesvandamme** [gc3-uzh-ch/elasticluster#575]
Dear all, 

After the 'elasticluster start mycluster' command, the setup ends with an error for the 'r : Install R MPI library with OpenMPI' task. Therefore my cluster is not ready to run the R code that I've written. Riccardo Murri is currently busy trying to fix this bug in the setup of the cluster. So if you're having the same problem: no worries, help is on the way!

Kind regards, 
Charles

**Comment by JohnAntonusMaximus** [comment]
Hi! This isn't closed, I'm repeatedly getting this error on Google Cloud. Here is the log output. I'm on latest for branch 'master':

```
TASK [r : Install R MPI library with OpenMPI] **********************************************************************************************************************************************************
task path: /home/elasticluster/share/playbooks/roles/r/tasks/rmpi.yml:12

fatal: [compute003]: FAILED! => {"changed": true, "cmd": "# load the MPI environment\n . /etc/profile.d/modules.sh\n export MODULEPATH=${MODULEPATH:-/etc/modulefiles}\n module load mpi\n # install Rmpi; note it is assumed that env vars MPI_HOME etc. are\n # defined in the module file above\n R --slave --no-save --no-restore-history <<-__EOF__\n if ('Rmpi' %in% installed.packages()[,'Package']) {\n # nothing to do, successful exit\n quit(save='no', status=0);\n } else {\n install.packages('Rmpi',\n repos='https://cloud.r-project.org/',\n configure.args=c('--with-Rmpi-type=OPENMPI',\n '--with-mpi=${MPI_HOME}',\n '--with-Rmpi-include=${MPI_INCLUDE}',\n '--with-Rmpi-libpath=${MPI_LIB}'));\n # 'install.packages' may fail but does not throw an error\n # condition so check whether the package was actually\n # installed and exit accordingly\n if (require('Rmpi')) {\n # successful exit\n quit(save='no', status=0);\n } else {\n # error exit\n quit(save='no', status=1);\n };\n };\n __EOF__", "delta": "0:00:00.002770", "end": "2019-04-16 20:07:47.977813", "msg": "non-zero return code", "rc": 2, "start": "2019-04-16 20:07:47.975043", "stderr": "/bin/sh: 2: .: Can't open /etc/profile.d/modules.sh", "stderr_lines": ["/bin/sh: 2: .: Can't open /etc/profile.d/modules.sh"], "stdout": "", "stdout_lines": []}
fatal: [frontend001]: FAILED! => {"changed": true, "cmd": "# load the MPI environment\n . /etc/profile.d/modules.sh\n export MODULEPATH=${MODULEPATH:-/etc/modulefiles}\n module load mpi\n # install Rmpi; note it is assumed that env vars MPI_HOME etc. are\n # defined in the module file above\n R --slave --no-save --no-restore-history <<-__EOF__\n if ('Rmpi' %in% installed.packages()[,'Package']) {\n # nothing to do, successful exit\n quit(save='no', status=0);\n } else {\n install.packages('Rmpi',\n repos='https://cloud.r-project.org/',\n configure.args=c('--with-Rmpi-type=OPENMPI',\n '--with-mpi=${MPI_HOME}',\n '--with-Rmpi-include=${MPI_INCLUDE}',\n '--with-Rmpi-libpath=${MPI_LIB}'));\n # 'install.packages' may fail but does not throw an error\n # condition so check whether the package was actually\n # installed and exit accordingly\n if (require('Rmpi')) {\n # successful exit\n quit(save='no', status=0);\n } else {\n # error exit\n quit(save='no', status=1);\n };\n };\n __EOF__", "delta": "0:00:00.002814", "end": "2019-04-16 20:07:48.028205", "msg": "non-zero return code", "rc": 2, "start": "2019-04-16 20:07:48.025391", "stderr": "/bin/sh: 2: .: Can't open /etc/profile.d/modules.sh", "stderr_lines": ["/bin/sh: 2: .: Can't open /etc/profile.d/modules.sh"], "stdout": "", "stdout_lines": []}

NO MORE HOSTS LEFT *************************************************************************************************************************************************************************************

PLAY RECAP *********************************************************************************************************************************************************************************************
compute001                 : ok=146  changed=74   unreachable=0    failed=1   
compute002                 : ok=146  changed=74   unreachable=0    failed=1   
compute003                 : ok=146  changed=74   unreachable=0    failed=1   
frontend001                : ok=158  changed=87   unreachable=0    failed=1   

2019-04-16 16:07:48 7b218b6f8157 gc3.elasticluster[1] ERROR Command `ansible-playbook --private-key=/home/slurm_user/.ssh/google_compute_engine /home/elasticluster/share/playbooks/main.yml --inventory=/home/slurm_user/.elasticluster/storage/slurmcluster.inventory --become --become-user=root -vv -e @extra_vars.yml` failed with exit code 2.
```

**Comment by riccardomurri** [comment]
@JohnAntonusMaximus can you share the config you're using to build the cluster? (Watch out and remove all access codes and passwords!)

Also: are you using the Dockerized version of ElastiCluster (`elasticluster.sh`) or have you installed the Python code from a GitHub clone?

**Comment by JohnAntonusMaximus** [comment]
Hey Riccardo, thank you for the quick response, below is my config file. Using the Elasticluster.sh file, and its pulling the latest docker image. Trying switching to a CentOS 2017 image but that through a whole bunch of other errors, so I switched back to debian. It appears the modules.sh file is missing from the compute nodes when I ssh into them and cd into the **/etc/profile.d/** folder:

```
[cloud/google]
provider=google
gce_client_id=___REMOVED_FOR_SECURITY____
gce_client_secret=____REMOVED_FOR_SECURITY_____
gce_project_id=slurm-project-1
noauth_local_webserver=True
zone=us-east1-c

[login/google]
image_user=john_r
image_user_sudo=root
image_sudo=True
user_key_name=elasticluster
user_key_private=~/.ssh/google_compute_engine
user_key_public=~/.ssh/google_compute_engine.pub

[setup/ansible]
ansible_forks=20
ansible_timeout=300

[setup/slurm]
provider=ansible
frontend_groups=slurm_master,r
compute_groups=slurm_worker,r

[cluster/slurmcluster]
cloud=google
login=google
setup=slurm
security_group=default
image_id=debian-9-stretch-v20190326
flavor=n1-standard-2
frontend_nodes=1
compute_nodes=3
ssh_to=frontend
boot_disk_size=50

[cluster/slurmcluster/frontend]
boot_disk_type=pd-ssd
boot_disk_size=1000
```

**Comment by riccardomurri** [comment]
I think the bug was reintroduced by the recent modifications to the Lmod role.

Now it's fixed (again) in commits a106c7d7 through a2baa95f.

**Comment by riccardomurri** [comment]
@JohnAntonusMaximus regarding issues with CentOS: can you please open a different issue about that?  Please post the failing config and exact error message.
